hook.c

```diff
@@ -73,6 +74,84 @@ static struct HookList Hooks = TAILQ_HEAD_INITIALIZER(Hooks);
 static struct HashTable *IdxFmtHooks = NULL;
 static HookFlags current_hook_type = MUTT_HOOK_NO_FLAGS;
 
+/**
+ * mutt_parse_charset_hook - Parse the 'charset-hook' commands - Implements Command::parse() - @ingroup command_parse
+ */
+enum CommandResult mutt_parse_charset_hook(struct Buffer *buf, struct Buffer *s,
+                                           intptr_t data, struct Buffer *err)
+{
+  struct Buffer *alias = mutt_buffer_pool_get();
+  struct Buffer *charset = mutt_buffer_pool_get();
+
+  int rc = MUTT_CMD_ERROR;
+  int retval = 0;
+
+  retval += mutt_extract_token(alias, s, MUTT_TOKEN_NO_FLAGS);
+  retval += mutt_extract_token(charset, s, MUTT_TOKEN_NO_FLAGS);
+
+  if (mutt_buffer_is_empty(alias))
+  {
+    mutt_buffer_printf(err, _("%s: too few arguments"), buf->data);
+    rc = MUTT_CMD_WARNING;
+  }
+  else if (mutt_buffer_is_empty(charset))
+  {
+    mutt_buffer_printf(err, _("%s: too few arguments"), buf->data);
+    rc = MUTT_CMD_WARNING;
+  }
+  else if (mutt_ch_lookup_add(MUTT_LOOKUP_CHARSET, mutt_buffer_string(alias),
+                              mutt_buffer_string(charset), err))
+  {
+    rc = MUTT_CMD_SUCCESS;
+  }
+
+  rc = (retval == 0) && (rc == MUTT_CMD_SUCCESS) ? MUTT_CMD_SUCCESS : rc;
+
+  mutt_buffer_pool_release(&alias);
+  mutt_buffer_pool_release(&charset);
+
+  return rc;
+}
+
+/**
+ * mutt_parse_iconv_hook - Parse the 'iconv-hook' commands - Implements Command::parse() - @ingroup command_parse
+ */
+enum CommandResult mutt_parse_iconv_hook(struct Buffer *buf, struct Buffer *s,
+                                         intptr_t data, struct Buffer *err)
+{
+  struct Buffer *alias = mutt_buffer_pool_get();
+  struct Buffer *charset = mutt_buffer_pool_get();
+
+  int rc = MUTT_CMD_ERROR;
+  int retval = 0;
+
+  retval += mutt_extract_token(alias, s, MUTT_TOKEN_NO_FLAGS);
+  retval += mutt_extract_token(charset, s, MUTT_TOKEN_NO_FLAGS);
+
+  if (mutt_buffer_is_empty(alias))
+  {
+    mutt_buffer_printf(err, _("%s: too few arguments"), buf->data);
+    rc = MUTT_CMD_WARNING;
+  }
+  else if (mutt_buffer_is_empty(charset))
+  {
+    mutt_buffer_printf(err, _("%s: too few arguments"), buf->data);
+    rc = MUTT_CMD_WARNING;
+  }
+  else if (mutt_ch_lookup_add(MUTT_LOOKUP_ICONV, mutt_buffer_string(alias),
+                              mutt_buffer_string(charset), err))
+  {
+    rc = MUTT_CMD_SUCCESS;
+  }
+
+  rc = (retval == 0) && (rc == MUTT_CMD_SUCCESS) ? MUTT_CMD_SUCCESS : MUTT_CMD_ERROR;
```

Comment:
Pardon? :-)
Is it possible to make this more legible?

Reply:
Heheh. I'm also not happy with this code.

It was partially driven by the confusion I got from not getting an error after calling mutt_extract_token(), even though no more text was available in the buffer.
I just wanted to retain the error message in case of refactoring of the mutt_extract_token(), if it might have shown up as necessary. But maybe I'm not familiar enough with the concept of these parser functions.

The documentation does not clarify it: ```mutt_extract_token - Extract one token from a string```.
```@retval  0 Success``` and ```@retval -1 Error```.

For me extracting a token means to get something back. But if it's not possible, because there is nothing left to be tokenized, I would expect an error of some kind. It seems error for this function means a more radical issue than just not extracting something. Maybe there should be more than two cases: success, parse-error and more-fundamental error (like NULL ptr or such stuff).

The other thing I had questions on, was: when to return ```MUTT_CMD_ERROR``` and when ```MUTT_CMD_WARNING```.
Would an eror from ```mutt_extract_token()``` justify the error case? I woul think so.
But would parse error also justify that? I would rather think the warning case would be enough.
But there might be cases or philosophies that would also vote for the error case here.
For example "fail early": in case of misconfigured rc-file don't start up and insist on correct config.

So a lot of unanswered questions. that lead to that "just in case" code.

Given the behaviour of ```mutt_extract_token()``` as it is, I can rewrite my code and look out for a better solution.

Reply:
> not getting an error after calling mutt_extract_token(),
> even though no more text was available in the buffer

Hmm...
I'm not against fixing the behaviour, but it's used in ~90 places, so we'd need a good plan first.

> when to return MUTT_CMD_ERROR and when MUTT_CMD_WARNING

Mutt doesn't have a `mutt_warning()`.
I introduced it do reduce the number of "errors" shown to the user.

- Warnings: When we can tell the user what they did wrong,
  so they can avoid the problem next time

- Errors: When the user couldn't have avoided the problem

> Would an eror from mutt_extract_token() justify the error case?

Probably, yes.
It's most likely triggered if there's a mistake in the code.